src/dune_rpc_impl/server.ml

```diff
@@ -329,25 +335,33 @@ let handler (t : _ t Fdecl.t) handle : 'build_arg Dune_rpc_server.Handler.t =
     Handler.declare_notification rpc Procedures.Server_side.log
   in
   let () = Handler.implement_request rpc Procedures.Public.ping (fun _ -> Fiber.return) in
-  let () =
-    let build _session targets =
+  let implement_request_pending_action decl ~f =
+    let handler _session input =
       let server = Fdecl.get t in
-      let ivar = Fiber.Ivar.create () in
-      let targets = List.map targets ~f:server.parse_build_arg in
-      let* () = Job_queue.write server.pending_build_jobs (targets, ivar) in
-      let+ build_outcome = Fiber.Ivar.read ivar in
+      let outcome = Fiber.Ivar.create () in
+      let* () = Job_queue.write server.pending_jobs { kind = f input; outcome } in
+      let+ build_outcome = Fiber.Ivar.read outcome in
       match (build_outcome : Build_outcome.t) with
       | Success -> Build_outcome_with_diagnostics.Success
       | Failure -> Failure (get_current_diagnostic_errors ())
     in
-    Handler.implement_request rpc Decl.build build
+    Handler.implement_request rpc decl handler
+  in
+  let () =
+    implement_request_pending_action Decl.build ~f:(fun targets ->
+      let server = Fdecl.get t in
+      let targets = List.map targets ~f:server.parse_build_arg in
+      Build targets)
+  in
+  let () =
+    implement_request_pending_action Decl.runtest ~f:(fun paths -> Runtest paths)
   in
   let () =
     let rec cancel_pending_jobs () =
-      match Job_queue.pop_internal (Fdecl.get t).pending_build_jobs with
+      match Job_queue.pop_internal (Fdecl.get t).pending_jobs with
       | None -> Fiber.return ()
-      | Some (_, job) ->
-        let* () = Fiber.Ivar.fill job Build_outcome.Failure in
+      | Some pending_action ->
```

Comment:
Match on just `{outcome; kind=_}` to give the compiler chances to complain. Or `{outcome; _}` to let the user know we won't use other values.